### PR TITLE
prov/efa: Rework zcpy recv setting

### PR DESCRIPTION
**Pull request by shijin-aws**
This patch refactor the efa_rdm_ep_set_use_zcpy_rx function for better readability and logging. It
also removes the requirement for FI_MSG_PREFIX

### CODE REVIEW

**Review by darrylabbate — COMMENTED** [review]
Good change. Let's use e.g. "Zero-copy receive protocol" over "zcpy rx" for logging, documentation, etc.

Also, we seem to be missing documentation for `FI_EFA_USE_ZCPY_RX` in `fi_efa(7)`

**Comment by shijin-aws** [comment]
> Good change. Let's use e.g. "Zero-copy receive protocol" over "zcpy rx" for logging, documentation, etc.
> 
> Also, we seem to be missing documentation for `FI_EFA_USE_ZCPY_RX` in `fi_efa(7)`

Good point, updated in the latest revision

File: man/fi_efa.7.md [hunk 1]
```diff
@@ -309,6 +309,14 @@ Using huge page memory has a small performance advantage, but can
 cause system to run out of huge page memory. By default, EFA provider
 will use huge page unless FI_EFA_FORK_SAFE is set to 1/on/true.
 
+*FI_EFA_USE_ZCPY_RX*
+: Enables the use of application's receive buffers in place of bounce-buffers when feasible.
+(Default: 1). Setting this environment variable to 0 can disable this feature.
+However, setting this variable to 1 is not sufficient to finally enable this feature because
+it has other requirements, see
+https://github.com/ofiwg/libfabric/blob/main/prov/efa/docs/efa_rdm_protocol_v4.md
+for details.
```
Comment (darrylabbate) [inline man/fi_efa.7.md:318]:
```suggestion
Explicitly setting this variable to 1 does not guarantee this feature
due to other requirements. See
https://github.com/ofiwg/libfabric/blob/main/prov/efa/docs/efa_rdm_protocol_v4.md
for details.
```

File: prov/efa/src/rdm/efa_rdm_ep_fiops.c [hunk 2]
```diff
@@ -394,16 +394,43 @@ static struct fi_ops efa_rdm_ep_base_ops = {
...
+	/* Unsupported capabilities */
+	if (ep->base_ep.util_ep.caps & unsupported_caps) {
+		EFA_INFO(FI_LOG_EP_CTRL, "Unsupported capabilities, disable zero-copy receive protocol\n");
+		ep->use_zcpy_rx = false;
+		goto out;
+	}
+
+	/* Max msg size is too large, turn off zcpy recv */
+	if (ep->max_msg_size > ep->mtu_size - ep->user_info->ep_attr->msg_prefix_size) {
+		EFA_INFO(FI_LOG_EP_CTRL, "max_msg_size %lu is greater than the mtu size limit: %lu, turn off zero-copy receive protocol\n",
```
Comment (darrylabbate) [inline prov/efa/src/rdm/efa_rdm_ep_fiops.c:417]:
```suggestion
		EFA_INFO(FI_LOG_EP_CTRL, "max_msg_size (%zu) is greater than the mtu size limit: %zu. Zero-copy receive protocol will be disabled.\n",
```

File: prov/efa/src/rdm/efa_rdm_ep_fiops.c [hunk 3]
```diff
@@ -394,16 +394,43 @@ static struct fi_ops efa_rdm_ep_base_ops = {
...
+	/* User requests to turn off zcpy recv */
+	if (!efa_env.use_zcpy_rx) {
+		EFA_INFO(FI_LOG_EP_CTRL, "User disables zero-copy receive protocol via environment\n");
+		ep->use_zcpy_rx = false;
+		goto out;
+	}
+
+	/* Unsupported capabilities */
+	if (ep->base_ep.util_ep.caps & unsupported_caps) {
+		EFA_INFO(FI_LOG_EP_CTRL, "Unsupported capabilities, disable zero-copy receive protocol\n");
```
Comment (darrylabbate) [inline prov/efa/src/rdm/efa_rdm_ep_fiops.c:410]:
```suggestion
		EFA_INFO(FI_LOG_EP_CTRL, "Unsupported capabilities, zero-copy receive protocol will be disabled\n");
```

File: prov/efa/src/rdm/efa_rdm_ep_fiops.c [hunk 4]
```diff
@@ -394,16 +394,43 @@ static struct fi_ops efa_rdm_ep_base_ops = {
...
+	if (ep->max_msg_size > ep->mtu_size - ep->user_info->ep_attr->msg_prefix_size) {
+		EFA_INFO(FI_LOG_EP_CTRL, "max_msg_size %lu is greater than the mtu size limit: %lu, turn off zero-copy receive protocol\n",
+			ep->max_msg_size, ep->mtu_size - ep->user_info->ep_attr->msg_prefix_size);
+		ep->use_zcpy_rx = false;
+		goto out;
+	}
+
+	/* If app needs sas ordering, turn off zcpy recv */
+	if (efa_rdm_ep_need_sas(ep)) {
+		EFA_INFO(FI_LOG_EP_CTRL, "FI_ORDER_SAS is requested, disable zero-copy receive protocol\n");
```
Comment (darrylabbate) [inline prov/efa/src/rdm/efa_rdm_ep_fiops.c:425]:
```suggestion
		EFA_INFO(FI_LOG_EP_CTRL, "FI_ORDER_SAS is requested, zero-copy receive protocol will be disabled\n");
```

**Review by darrylabbate — APPROVED** [review]
Just nits

**Comment by shijin-aws** [comment]
> Just nits

Thank you ! Updated all

**Pull request merged**